File: app/src/main/java/com/mstc/mstcapp/ViewResourceActivity.java
package com.mstc.mstcapp;

import android.content.Context;
import android.os.Bundle;
import android.view.MenuItem;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.appbar.CollapsingToolbarLayout;
import com.google.android.material.tabs.TabLayout;
import com.mstc.mstcapp.model.DomainModel;
import com.mstc.mstcapp.ui.resources.ViewPagerAdapter;

import java.util.Objects;

public class ViewResourceActivity extends AppCompatActivity {
    private final Context context = this;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DomainModel domain = (DomainModel) getIntent().getSerializableExtra("domain");

        if (domain == null) {
            Toast.makeText(context, "Could not load domain! Please try again...", Toast.LENGTH_SHORT).show();
            finish();
        } else {
            setTheme(domain.getColor());

            setContentView(R.layout.activity_view_resource);
            Toolbar toolbar = findViewById(R.id.toolbar);
            setSupportActionBar(toolbar);
            Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setHomeAsUpIndicator(ContextCompat.getDrawable(context, R.drawable.ic_back));

            TextView toolbar_title = findViewById(R.id.toolbar_title);
            toolbar_title.setText(domain.getDomain().toUpperCase());

            TextView toolbar_description = findViewById(R.id.toolbar_description);
            toolbar_description.setText(R.string.choose_domain_helper_text);

            ImageView toolbar_image = findViewById(R.id.toolbar_image);
            toolbar_image.setImageDrawable(ContextCompat.getDrawable(context, domain.getDrawable()));

            TabLayout tabLayout = findViewById(R.id.tabLayout);
            ViewPager viewPager = findViewById(R.id.viewPager);
            ViewPagerAdapter viewPagerAdapter = new ViewPagerAdapter(getSupportFragmentManager(), domain.getDomain().toLowerCase());
            viewPager.setAdapter(viewPagerAdapter);
            tabLayout.setupWithViewPager(viewPager);
            CollapsingToolbarLayout collapsingToolbarLayout = findViewById(R.id.collapsingToolbarLayout);
            collapsingToolbarLayout.setTitle(domain.getDomain().toUpperCase());
        }

    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home)
            finish();
        return true;
    }
}

File: app/src/main/java/com/mstc/mstcapp/ViewResourceActivity.kt
package com.mstc.mstcapp

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.view.animation.AnimationUtils
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.google.android.material.appbar.AppBarLayout
import com.mstc.mstcapp.databinding.ActivityViewResourceBinding
import com.mstc.mstcapp.model.Domain
import com.mstc.mstcapp.ui.resources.ViewPagerAdapter
import java.util.*
import kotlin.math.abs

private const val TAG = "ViewResourceActivity"

private enum class State {
    EXPANDED, COLLAPSED, IDLE
}

class ViewResourceActivity : AppCompatActivity() {
    private val context: Context = this
    lateinit var binding: ActivityViewResourceBinding
    private lateinit var domainModel: Domain
    private var state: State? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        domainModel = intent.getSerializableExtra("domain") as Domain
        setTheme(domainModel.style)
        binding = ActivityViewResourceBinding.inflate(layoutInflater)
        setContentView(binding.root)
        binding.apply {
            setSupportActionBar(toolbar)
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
            supportActionBar?.setHomeAsUpIndicator(
                ContextCompat.getDrawable(
                    context,
                    R.drawable.ic_back
                )
            )
            toolbarTitle.text = domainModel.domain.uppercase(Locale.getDefault())
            toolbarImage.setImageDrawable(
                ContextCompat.getDrawable(
                    context,
                    domainModel.drawable
                )
            )
            Log.i(TAG, "onCreate: ${toolbar.title}  , ${collapsingToolbarLayout.title}")
            appBarLayout.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { _, verticalOffset ->
                when {
                    verticalOffset == 0 -> {
                        if (state != State.EXPANDED) {
                            Log.d(TAG, "Expanded")
                            toolbarImage.animation = AnimationUtils.loadAnimation(
                                this@ViewResourceActivity,
                                android.R.anim.fade_in
                            )
                            toolbarImage.postOnAnimation {
                                toolbarImage.visibility = View.VISIBLE
                                toolbarTitle.visibility = View.VISIBLE
                                toolbarDescription.visibility = View.VISIBLE
                            }
                        }
                        state = State.EXPANDED
                    }
                    abs(verticalOffset) >= appBarLayout.totalScrollRange -> {
                        if (state != State.COLLAPSED) {
                            Log.d(TAG, "Collapsed")
                        }
                        state = State.COLLAPSED
                    }
                    else -> {
                        if (state != State.IDLE) {
                            Log.d(TAG, "Idle")
                            toolbarImage.animation = AnimationUtils.loadAnimation(
                                this@ViewResourceActivity,
                                android.R.anim.fade_out
                            )
                            toolbarImage.postOnAnimation {
                                toolbarImage.visibility = View.INVISIBLE
                                toolbarTitle.visibility = View.INVISIBLE
                                toolbarDescription.visibility = View.INVISIBLE
                            }
                        }
                        state = State.IDLE
                    }
                }
            })
        }
        showData()
    }

    private fun showData() {
        val viewPagerAdapter =
            ViewPagerAdapter(
                supportFragmentManager,
                domainModel.domain.lowercase(Locale.getDefault())
            )
        binding.apply {
            viewPager.adapter = viewPagerAdapter
            tabLayout.setupWithViewPager(binding.viewPager)
            collapsingToolbarLayout.title = domainModel.domain.uppercase(Locale.getDefault())
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        finish()
        return true
    }

    override fun onBackPressed() {
        when (binding.viewPager.currentItem) {
            0 -> finish()
            else -> binding.viewPager.currentItem = 0
        }
    }

    override fun onStart() {
        super.onStart()
        overridePendingTransition(R.anim.slide_in_right, android.R.anim.fade_out)
    }

    override fun finish() {
        super.finish()
        overridePendingTransition(0, R.anim.slide_out_left)
    }
}